Add supports.spacing.blockGap to block.json schema

t-hamano opened the pull request:
## What?

This PR adds `supports.spacing.blockGap` property to the `block.json` schema.

![block](https://user-images.githubusercontent.com/54422211/215116173-8242695c-19b1-4925-a9e8-89ae20ce4354.png)

## Why?
To receive input support by auto-completion and to be informed of incorrect value input by validation.

## How?

This new property accepts the following two types:

### Boolean

Opt-in horizontal vertical common gap support:

```json
{
	"supports": {
		"spacing": {
			"blockGap": true
		}
	}
}
```

### Array

Opt-in for separate horizontal and vertical controls ~or one of them~:

```json
{
	"supports": {
		"spacing": {
			"blockGap": [ "horizontal", "vertical" ]
		}
	}
}
```

### About unsupported type

In the column block, the `blockGap` property is an object type and has `__experimentalDefault` property.

```json
{
	"supports": {
		"spacing": {
			"blockGap": {
				"__experimentalDefault": "2em",
				"sides": [ "horizontal", "vertical" ]
			}
		}
	}
}
```

In this case, the `sides` property defines the direction to be supported. However, since experimental properties are not defined in the schema, there is no need to use the sides property. Therefore, I excluded the object type.

## Testing Instructions

Create a JSON file like the following, referencing the JSON schema defined by this PR:

```json
{
	"$schema": "https://raw.githubusercontent.com/WordPress/gutenberg/fix/block-json-schema-blockgap/schemas/json/block.json",
	"apiVersion": 2,
	"name": "core/test",
	"title": "Test",
	"supports": {
		"spacing": {
		}
	}
}
```
In the `settings.supports.spacing` property, confirm that the code editor assists with auto-completion and validation.

## Screenshots or screencast

https://user-images.githubusercontent.com/54422211/215121502-5003c63f-bd07-41e6-98ef-be05f97358ea.mp4

t-hamano committed:
Add supports.spacing.blockGap to block.json schema

andrewserong commented on schemas/json/block.json:
```diff
@@ -377,6 +377,21 @@
+									"type": "array",
+									"items": {
+										"type": "string",
+										"enum": [ "vertical", "horizontal" ]
```
I know you've already opened up a discussion about it in #47518 and #47523, but currently this array only supports including both of these values, it doesn't support only including one direction.

andrewserong reviewed (commented):
Thanks again for opening this one up, too @t-hamano! Since `blockGap` output is intrinsically tied to the layout block support implementation, it might be better to add this property to the schema at the same time as stabilising the layout API, otherwise it might sound like blocks can define `supports.spacing.blockGap` as `true` and then expect to have block spacing controls. However, this won't be possible for non-core blocks to do in a stable way until the layout support is stabilised. CC: @tellthemachines who's also looking at layout stabilisation tasks

t-hamano committed:
Add default array values

t-hamano committed:
Add description

t-hamano commented:
Based on [this discussion](https://github.com/WordPress/gutenberg/issues/47518#issuecomment-1409694415), I have added a `description` of the `blockGap` property that it does not support a single direction. At the same time, I have updated the screencast of the first comments in this PR.

Ideally, we should probably add a rule that expects both values to be specified in the array, but I couldn't find a way to define such a thing in the JSON schema.

t-hamano commented:
I have noticed that unit tests on the block schema fail.

```
FAIL test/integration/blocks-schema.test.js
  ● block.json schema › validates schema for `core/columns`

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: [{"instancePath": "/supports/spacing/blockGap", "keyword": "type", "message": "must be boolean", "params": {"type": "boolean"}, "schemaPath": "#/properties/supports/properties/spacing/properties/blockGap/oneOf/0/type"}, {"instancePath": "/supports/spacing/blockGap", "keyword": "type", "message": "must be array", "params": {"type": "array"}, "schemaPath": "#/properties/supports/properties/spacing/properties/blockGap/oneOf/1/type"}, {"instancePath": "/supports/spacing/blockGap", "keyword": "oneOf", "message": "must match exactly one schema in oneOf", "params": {"passingSchemas": null}, "schemaPath": "#/properties/supports/properties/spacing/properties/blockGap/oneOf"}]
```

This is because the column block has the following form, which is not defined in the schema:

```
{
		"spacing": {
			"blockGap": {
				"__experimentalDefault": "2em",
				"sides": [ "horizontal", "vertical" ]
			}
		}
}
```

Perhaps the schema might not need to be updated until the blockGap property is fully stabilized 😅

andrewserong commented:
> Perhaps the schema might not need to be updated until the blockGap property is fully stabilized 😅

Ah, good catch. Yes, might be worth holding off just for now. It's a great idea to stabilise it, though. Unfortunately I think we've probably just missed the cutoff for stabilising for 6.2, but let's see if we can get it stabilised for 6.3

t-hamano commented:
Yes, I would like to close this PR once and for all.
In the future, we can re-consider updating the schema when the API is stable 👍
